File: 348 Design Tic-Tac-To.java
/*
348. Design Tic-Tac-Toe
https://leetcode.com/problems/design-tic-tac-toe/

Design a Tic-tac-toe game that is played between two players on a n x n grid.

You may assume the following rules:

A move is guaranteed to be valid and is placed on an empty block.
Once a winning condition is reached, no more moves is allowed.
A player who succeeds in placing n of their marks in a horizontal, vertical, or diagonal row wins the game.
Example:
Given n = 3, assume that player 1 is "X" and player 2 is "O" in the board.

TicTacToe toe = new TicTacToe(3);

toe.move(0, 0, 1); -> Returns 0 (no one wins)
|X| | |
| | | |    // Player 1 makes a move at (0, 0).
| | | |

toe.move(0, 2, 2); -> Returns 0 (no one wins)
|X| |O|
| | | |    // Player 2 makes a move at (0, 2).
| | | |

toe.move(2, 2, 1); -> Returns 0 (no one wins)
|X| |O|
| | | |    // Player 1 makes a move at (2, 2).
| | |X|

toe.move(1, 1, 2); -> Returns 0 (no one wins)
|X| |O|
| |O| |    // Player 2 makes a move at (1, 1).
| | |X|

toe.move(2, 0, 1); -> Returns 0 (no one wins)
|X| |O|
| |O| |    // Player 1 makes a move at (2, 0).
|X| |X|

toe.move(1, 0, 2); -> Returns 0 (no one wins)
|X| |O|
|O|O| |    // Player 2 makes a move at (1, 0).
|X| |X|

toe.move(2, 1, 1); -> Returns 1 (player 1 wins)
|X| |O|
|O|O| |    // Player 1 makes a move at (2, 1).
|X|X|X|
Follow up:
Could you do better than O(n2) per move() operation?
*/
class TicTacToe {
    
    // The size is width * 2 +  2, row count + column count + 2 diagonals
    // counter[i] records the filled blocks diff between 2 players
    private int[] counter;
    private int width;
    
    /** Initialize your data structure here. */
    public TicTacToe(int n) {
        width = n;        
        counter = new int[n * 2 + 2];
    }
    
    /** Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins. */
    public int move(int row, int col, int player) {
        int block = (player == 1 ? 1 : -1);
        
        // Is current row filled ?
        if (updateAndJudge(row, block)) {
            return player;
        }
        
        // Is current column filled ?
        if (updateAndJudge(col + width, block)) {
            return player;
        }
        
        // Is top-left to bottom-right diagonal filled ?
        if (row == col && updateAndJudge(width * 2, block)) {
            return player;
        }
        
        // Is top-right to bottom-left diagonal filled ?
        if (row == width - 1 - col && updateAndJudge(width * 2 + 1, block)) {
            return player;
        }

        return 0;
    }
    
    /**
        @param row index of a row, a column or one of the 2 diag
        @param block 1 or -1
        @return true if there is a win or false if there is not yet a win
    */
    private boolean updateAndJudge(int index, int block) {
        counter[index] += block;
        return Math.abs(counter[index]) == width;
    }
}

/**
 * Your TicTacToe object will be instantiated and called as such:
 * TicTacToe obj = new TicTacToe(n);
 * int param_1 = obj.move(row,col,player);
 */

File: 348 Design Tic-Tac-To.ts
/*
348. Design Tic-Tac-Toe
https://leetcode.com/problems/design-tic-tac-toe/

Assume the following rules are for the tic-tac-toe game on an n x n board between two players:

    A move is guaranteed to be valid and is placed on an empty block.
    Once a winning condition is reached, no more moves are allowed.
    A player who succeeds in placing n of their marks in a horizontal, vertical, or diagonal row wins the game.

Implement the TicTacToe class:

    TicTacToe(int n) Initializes the object the size of the board n.
    int move(int row, int col, int player) Indicates that the player with id player plays at the cell (row, col) of the board.
    The move is guaranteed to be a valid move, and the two players alternate in making moves. Return
        0 if there is no winner after the move,
        1 if player 1 is the winner after the move, or
        2 if player 2 is the winner after the move.

Example 1:

Input
["TicTacToe", "move", "move", "move", "move", "move", "move", "move"]
[[3], [0, 0, 1], [0, 2, 2], [2, 2, 1], [1, 1, 2], [2, 0, 1], [1, 0, 2], [2, 1, 1]]
Output
[null, 0, 0, 0, 0, 0, 0, 1]

Explanation
TicTacToe ticTacToe = new TicTacToe(3);
Assume that player 1 is "X" and player 2 is "O" in the board.
ticTacToe.move(0, 0, 1); // return 0 (no one wins)
|X| | |
| | | |    // Player 1 makes a move at (0, 0).
| | | |

ticTacToe.move(0, 2, 2); // return 0 (no one wins)
|X| |O|
| | | |    // Player 2 makes a move at (0, 2).
| | | |

ticTacToe.move(2, 2, 1); // return 0 (no one wins)
|X| |O|
| | | |    // Player 1 makes a move at (2, 2).
| | |X|

ticTacToe.move(1, 1, 2); // return 0 (no one wins)
|X| |O|
| |O| |    // Player 2 makes a move at (1, 1).
| | |X|

ticTacToe.move(2, 0, 1); // return 0 (no one wins)
|X| |O|
| |O| |    // Player 1 makes a move at (2, 0).
|X| |X|

ticTacToe.move(1, 0, 2); // return 0 (no one wins)
|X| |O|
|O|O| |    // Player 2 makes a move at (1, 0).
|X| |X|

ticTacToe.move(2, 1, 1); // return 1 (player 1 wins)
|X| |O|
|O|O| |    // Player 1 makes a move at (2, 1).
|X|X|X|

Constraints:
    2 <= n <= 100
    player is 1 or 2.
    0 <= row, col < n
    (row, col) are unique for each different call to move.
    At most n2 calls will be made to move.

Follow-up: Could you do better than O(n2) per move() operation?
*/
class TicTacToe {
    private width: number;
    // Save the record of the filling on each row.
    // If rowFill[i] = width or -width, a winner.
    private rowFills: number[];
    private columnFills: number[];
    private diagonalFills: number[];

    constructor(n: number) {
        this.width = n;
        this.rowFills = new Array(n).fill(0);
        this.columnFills = new Array(n).fill(0);
        this.diagonalFills = [0, 0];
    }

    move(row: number, col: number, player: number): number {
        const block = player == 1 ? 1 : -1;
        // Row
        this.rowFills[row] += block;
        if (Math.abs(this.rowFills[row]) === this.width) {
            return player;
        }
        // Column
        this.columnFills[col] += block;
        if (Math.abs(this.columnFills[col]) === this.width) {
            return player;
        }
        // Top-left to bottom-right diagonal
        if (row === col) {
            this.diagonalFills[0] += block;
            if (Math.abs(this.diagonalFills[0]) === this.width) {
                return player;
            }
        }
        // Top-right to bottom-left diagonal
        if (row === this.width - 1 - col) {
            this.diagonalFills[1] += block;
            if (Math.abs(this.diagonalFills[1]) === this.width) {
                return player;
            }
        }
        return 0;
    }
}
